pkg/hive/hive.go

```diff
@@ -0,0 +1,297 @@
+	lifecycle                 lifecycleProxy
+	startTimeout, stopTimeout time.Duration
+	viper                     *viper.Viper
+}
+
+// New returns a new hive that can be run, or inspected.
+// The command-line flags from the cells are registered as part of this.
+//
+// The object graph is not constructed until methods of the hive are
+// invoked.
+func New(v *viper.Viper, flags *pflag.FlagSet, cells ...*Cell) *Hive {
+	h := &Hive{
+		envPrefix:    defaultEnvPrefix,
+		fxLogger:     logging.NewFxLogger(log),
+		cells:        cells,
+		viper:        v,
+		flags:        pflag.NewFlagSet("", pflag.ContinueOnError),
+		startTimeout: defaultStartTimeout,
+		stopTimeout:  defaultStopTimeout,
+	}
+	if err := h.registerFlags(flags); err != nil {
+		log.Fatal(err)
+	}
+	return h
+}
+
+func (h *Hive) SetTimeouts(start, stop time.Duration) {
+	h.startTimeout, h.stopTimeout = start, stop
+}
+
+func (h *Hive) SetEnvPrefix(prefix string) {
+	h.envPrefix = prefix
+}
+
+// Run populates the cell configurations and runs the hive cells.
+// If an error occurs when populating cell configurations this
+// method panics.
+// Interrupt signal will cause the hive to stop.
+func (h *Hive) Run() {
+	if h.app == nil {
+		if err := h.createApp(); err != nil {
+			log.Fatalf("populate: %s", err)
+		}
+	}
+	h.app.Run()
+}
+
+func (h *Hive) Start(ctx context.Context) error {
+	if h.app == nil {
+		if err := h.createApp(); err != nil {
+			log.Fatalf("populate: %s", err)
+		}
+	}
+	return h.app.Start(ctx)
+}
+
+func (h *Hive) Stop(ctx context.Context) error {
+	if h.app == nil {
+		return errors.New("Hive was not started")
+	}
+	return h.app.Stop(ctx)
+}
+
+func (h *Hive) PrintObjects() {
+	if err := h.createApp(); err != nil {
+		log.Fatal(err)
+	}
+	h.fxLogger.PrintObjects()
+
+	fmt.Printf("Start hooks:\n\n")
+	for _, hook := range h.lifecycle.hooks {
+		if hook.OnStart == nil {
+			continue
+		}
+		fmt.Printf("  • %s\n", funcNameAndLocation(hook.OnStart))
+	}
+
+	fmt.Printf("\nStop hooks:\n\n")
+	for i := len(h.lifecycle.hooks) - 1; i >= 0; i-- {
+		hook := h.lifecycle.hooks[i]
+		if hook.OnStop == nil {
+			continue
+		}
+		fmt.Printf("  • %s\n", funcNameAndLocation(hook.OnStop))
+	}
+}
+
+func (h *Hive) PrintDotGraph() {
+	if err := h.createApp(); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Print(h.dotGraph)
+}
+
+// getEnvName returns the environment variable to be used for the given option name.
+func (h *Hive) getEnvName(option string) string {
+	under := strings.Replace(option, "-", "_", -1)
+	upper := strings.ToUpper(under)
+	return h.envPrefix + upper
+}
+
+// registerCells registers the command-line flags from all the cells.
+func (h *Hive) registerFlags(parent *pflag.FlagSet) error {
+	for _, cell := range h.cells {
+		cell.registerFlags(h.flags)
+	}
+	var err error
+	h.flags.VisitAll(func(f *pflag.Flag) {
+		if err != nil {
+			return
+		}
+		if parent.Lookup(f.Name) != nil {
+			err = fmt.Errorf("Error registering flags: '%s' already registered", f.Name)
```

Comment:
nit: capitalized error string (`error strings should not be capitalized (ST1005) go-staticcheck`)

Reply:
Hm, odd that `make lint` doesn't catch this. Do we have some check for this or was it just your editor setup that catches it?

Reply:
Running `make golangci-lint` gives nothing, so it is just my editor setting (I also can find some other cases like that, even if they are very few).

Reply:
Fixed.